sql/server.go

```diff
@@ -186,13 +211,32 @@ func (s *Server) exec(req driver.Request) (resp driver.Response, err error) {
 			return
 		}
 		var plan planNode
-		if err = s.db.Txn(func(txn *client.Txn) error {
-			planMaker.txn = txn
+		if _, ok := stmt.(*parser.BeginTransaction); ok {
+			if planMaker.txn != nil {
+				err = errBeginDuringOpenTxn
+				return
+			}
+			// Start a transaction here and not in planMaker because planMaker
+			// doesn't have access to the db object.
+			planMaker.txn = s.db.BeginTxn()
+		}
+
+		// If there is a pending transaction.
+		if planMaker.txn != nil {
+			// Run in transaction planMaker.txn
 			plan, err = planMaker.makePlan(stmt)
-			planMaker.txn = nil
-			return err
-		}); err != nil {
-			return
+			if err != nil {
+				return
+			}
+		} else {
+			if err = s.db.Txn(func(txn *client.Txn) error {
```

Comment:
Add a comment here that we take advantage of the `db.Txn` retry logic for auto-committed transactions. I was going to propose that you unify this else-block with the if-block above, but then realized the retry logic is important. I bet someone else will make the same mistake.